### PR TITLE
Kafka integration

### PR DESCRIPTION
**Pull request by nikofil**
Add integration with Kafka for publishing events in a similar fashion as ZeroMQ.

#### Questions

- [X] Does it require a DB change?
- [ ] Are you using it in production?
- [ ] Does it require a change in the API (PyMISP for example)?

#### Release Type:
- [ ] Major
- [X] Minor
- [ ] Patch

### CODE REVIEW

File: app/Model/AppModel.php [hunk 1]
```diff
@@ -1506,6 +1511,30 @@ public function setupRedis()
...
+            $this->loadKafkaPubTool();
+        }
+        return $this->loadedKafkaPubTool;
+    }
+
+    public function loadKafkaPubTool()
+    {
+        App::uses('KafkaPubTool', 'Tools');
+        $kafkaPubTool = new KafkaPubTool();
+        $kafkaConf = array(
```
Comment (liviuvalsan) [inline app/Model/AppModel.php]: This Kafka config array could be added directly into the config file.

File: app/Model/Attribute.php [hunk 2]
```diff
@@ -666,7 +668,14 @@ public function afterSave($created, $options = array())
                 if (Configure::read('Plugin.ZeroMQ_include_attachments') && $this->typeIsAttachment($attribute['Attribute']['type'])) {
```
Comment (liviuvalsan) [inline app/Model/Attribute.php]: This one needs to be handled separately for the Kafka case as well (`Plugin.Kafka_include_attachments`).

File: app/Model/Event.php [hunk 3]
```diff
@@ -483,6 +506,14 @@ public function afterSave($created, $options = array())
                 $pubSubTool->event_save($event, $created ? 'add' : 'edit');
             }
         }
+        $kafkaTopic = Configure::read('Plugin.Kafka_event_notifications_topic');
+        if (empty($this->data['Event']['unpublishAction']) && empty($this->data['Event']['skip_kafka']) && Configure::read('Plugin.Kafka_enable') && Configure::read('Plugin.Kafka_event_notifications_enable') && !empty($kafkaTopic)) {
+            $kafkaPubTool = $this->getKafkaPubTool();
+            $event = $this->quickFetchEvent($this->data['Event']['id']);
```
Comment (nikofil) [inline app/Model/Event.php]: also change this to the faster publishJson

File: app/Model/ShadowAttribute.php [hunk 4]
```diff
@@ -277,6 +277,15 @@ public function afterSave($created, $options = array())
         } else {
             $this->__afterSaveCorrelation($this->data['ShadowAttribute']);
         }
+        $kafkaTopic = Configure::read('Plugin.Kafka_shadow_attribute_notifications_topic');
+        if (Configure::read('Plugin.Kafka_enable') && Configure::read('Plugin.Kafka_shadow_attribute_notifications_enable') && !empty($kafkaTopic)) {
```
Comment (nikofil) [inline app/Model/ShadowAttribute.php]: add this one fn higher

File: app/View/Elements/genericElements/SideMenu/side_menu.ctp [hunk 5]
```diff
@@ -173,6 +173,16 @@
                             'message' => __('Are you sure you wish to republish the current event to the ZMQ channel?')
                         ));
                     }
+                    if (Configure::read('Plugin.Kafka_enable') &&
+                        Configure::read('Plugin.Kafka_event_publish_notifications_enable') &&
+                        Configure::read('Plugin.Kafka_event_publish_notifications_topic') &&
+                        $isAclKafka) {
+                        echo $this->element('/genericElements/SideMenu/side_menu_post_link', array(
```
Comment (nikofil) [inline app/View/Elements/genericElements/SideMenu/side_menu.ctp:180]: pub to _event by default, _publish_event iff published

**Pull request closed without merging**